sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java

```diff
@@ -636,4 +649,49 @@ public void visitPrimitiveTransform(Node node) {
       node.getTransform().validate(options);
     }
   }
+
+  private static class TransformToMessage implements Function<PTransform<?, ?>, String> {
+    @Override
+    public String apply(final PTransform<?, ?> transform) {
+      final Object representant;
+      if (ParDo.SingleOutput.class.isInstance(transform)) {
+        representant = ParDo.SingleOutput.class.cast(
+                transform).getFn();
+      } else if (ParDo.MultiOutput.class.isInstance(transform)) {
+        representant = ParDo.MultiOutput.class.cast(
+                transform).getFn();
+      } else {
+        representant = transform;
+      }
+      return "    - " + representant;
+    }
+  }
+
+  private static class UnstableNameToMessage implements Function<String, String> {
+    private final Multimap<String, PTransform<?, ?>> instances;
+
+    private UnstableNameToMessage(final Multimap<String, PTransform<?, ?>> instancePerName) {
+      this.instances = instancePerName;
+    }
+
+    @Override
+    public String apply(final String input) {
+      Collection<PTransform<?, ?>> values = null;
+      String currentName = input;
+      // strip the counter appended to the name,
+      // see uniquifyInternal().
+      // common example: ParDo(Anonymous) becoming ParDo(Anonymous)2
+      while (values == null && !currentName.isEmpty()) {
+        currentName = currentName.substring(0, currentName.length() - 1);
```

Comment:
What if the counter is greater than 10, and takes more than 1 digit?

In general this seems brittle: this is just extracting the original non-unique name of a transform, right? Can't we pass the offending PTransform's here directly somehow, e.g. at line 532 key things by the original (non-unique) name and traverse entries that have more than 1 transform? I think then you can also just remove the variable "unstableNames" because it's redundant w.r.t. the map.

Reply:
The loop will remove both digits normally.

Thought about removing it but then you need to browse the whole collection instead of just testing if the list is empty - twice. No strong opinion on it but thought that holding the direct information "conflict" was valuable.

Reply:
I think traversing the collection is much better than parsing digit-by-digit a string that we ourselves have formatted.

Reply:
Will change it tomorrow